src/components/command-bar/components/dialog/index.tsx

```diff
@@ -1,135 +1,32 @@
-import Text from 'components/text'
 import { CommandBarDialogFooterProps, CommandBarDialogProps } from './types'
+import CommandBarDialogHeader from './header'
+import CommandBarDialogFooter from './footer'
+import CommandBarDialogBody from './body'
 import s from './command-bar-dialog.module.css'
 
-const CommandBarDialogHeader = () => {
-	const { currentCommand, currentTags, removeTag } = useCommandBar()
-
-	return (
-		<div className={s.header}>
-			<div className={s.icon}>{currentCommand.icon}</div>
-			{currentTags.length > 0 ? (
-				<div className={s.headerTags}>
-					{currentTags.map((tag: CommandBarTag) => (
-						<Tag
-							key={tag.id}
-							text={tag.text}
-							onRemove={() => removeTag(tag.id)}
-						/>
-					))}
-				</div>
-			) : null}
-			<input
-				className={s.input}
-				placeholder={currentCommand.inputProps.placeholder}
-			/>
-			<div className={s.badges}>
-				<Badge
-					ariaLabel="Command key"
-					color="neutral"
-					icon={<IconCommand16 />}
-					size="small"
-					type="outlined"
-				/>
-				<Badge
-					ariaLabel="K key"
-					color="neutral"
-					size="small"
-					text="K"
-					type="outlined"
-				/>
-			</div>
-		</div>
-	)
-}
-
-const CommandBarDialogBody = () => {
-	const currentContentType = useCurrentContentType()
-
-	return (
-		<div className={s.body}>
-			<div>Current content type: {currentContentType}</div>
-		</div>
-	)
-}
-
-const CommandBarDialogFooter = () => {
-	return (
-		<div className={s.footer}>
-			<div className={s.footerLeft}>
-				<Badge color="neutral" text="Tab" type="outlined" size="small" />
-				<Text asElement="span" size={100} weight="regular">
-					or
-				</Text>
-				<div className={s.footerLeftArrowKeys}>
-					<Badge
-						ariaLabel="Up arrow key"
-						color="neutral"
-						icon={<IconArrowUp16 />}
-						type="outlined"
-						size="small"
-					/>
-					<Badge
-						ariaLabel="Down arrow key"
-						color="neutral"
-						icon={<IconArrowDown16 />}
-						type="outlined"
-						size="small"
-					/>
-				</div>
-				<Text asElement="span" size={100} weight="regular">
-					to navigate,
-				</Text>
-				<Text asElement="span" size={100} weight="regular">
-					to select,
-				</Text>
-				<Badge color="neutral" text="Esc" type="outlined" size="small" />
-				<Text asElement="span" size={100} weight="regular">
-					to exit
-				</Text>
-			</div>
-			<div>
-				<StandaloneLink
-					color="secondary"
-					href={FORM_URL}
-					icon={<IconExternalLink16 />}
-					iconPosition="trailing"
-					size="small"
-					text="Give Feedback"
-					openInNewTab
-				/>
-			</div>
-		</div>
-	)
-}
-
 const CommandBarDialog = ({
-	children,
 	isOpen = false,
 	onDismiss = () => null,
 }: CommandBarDialogProps) => {
+	// Get the current command from Command Bar state
+	const { currentCommand } = useCommandBar()
+
+	// Invoke currnet command's initial loading hook
+	const { useOnInitialLoad = () => null } = currentCommand
+	useOnInitialLoad()
```

Comment:
💭  This _might_ break the rules of hooks if/when someone switches to a command with a different `useOnInitialLoad` hook which itself has different calls to `useState`/`useEffect`/etc.

This feels like a potential justification for having commands support a custom `render` method, or custom body component, which would allow the command to run any arbitrary hooks / effects on render, without expanding the top-level `Command` interface.

Reply:
Is there a good way to test this?

How do you envision the `render` method working, and where do you see it being invoked? We need something that will allow all parts of the dialog to be prepared for UI, which is why I've placed this here.

Reply:
It could be tested by switching commands to something that has a different `useOnInitialLoad` hook.

I haven't given it much thought, but `render` would likely get called in `CommandBarDialogBody`:

```ts
  const { currentCommand } = useCommandBar()

	return (
		<div className={s.body}>
            {currentCommand.render()}
			<div>Current content type: {currentContentType}</div>
		</div>
	)
```

or...

```ts
  const { currentCommand } = useCommandBar()

	return (
		<div className={s.body}>
            {<currentCommand.Render />}
			<div>Current content type: {currentContentType}</div>
		</div>
	)
```

I don't think this would have any functional difference to what you're doing now. Effects don't block rendering in any way, so the sequencing would be the same.

```ts
const searchCommand = {
  // ...
  Render: () => {
    useOnInitialLoad()
    return (
      // search body
    )
  }
}
```